### ISSUE TITLE
Expose jax.core.eval_context() as part of the public API

### ISSUE THREAD

**Issue by FedericoV** [google/jax#7535]
Currently, this works:

```from functools import partial
import jax
import jax.numpy as jnp
from jax._src.custom_derivatives import closure_convert

idx = jnp.array([True, False])

def test_fcn(x):
  return x[idx]

@jax.jit
def f(x):
  g, args = closure_convert(test_fcn, x)
  return g(x, *args)

f(jnp.array([0., 1.]))
```

But this does not work unless you use jax.core.eval_context():

```
from functools import partial
import jax
import jax.numpy as jnp
from jax._src.custom_derivatives import closure_convert

idx = jnp.array([True, False])

def test_fcn(x):
  with jax.core.eval_context():
     bad_idx = idx == 0
  return x[bad_idx]

@jax.jit
def f(x):
  g, args = closure_convert(test_fcn, x)
  return g(x, *args)

f(jnp.array([0., 1.]))
```

The API should probably be renamed to something like - fold_at_compile_time or something similar, but, it's useful functionality to end users.

**Comment by mattjj** [comment]
Naming suggestions from @froystig:

> yeah, along those lines: const_{eval,fold}, const_{eval,fold}_only, {ensure,force}_compile_time, {ensure,force}_compile_time_eval

**Comment by froystig** [comment]
closed in #7987
